File: framework/languages/java/samples/java/GameQuest/Server/GameApi/src/main/java/systems/zlink/samples/gamequest/server/gameapi/handlers/NotifyQuestProgressHandler.java
package systems.zlink.samples.gamequest.server.gameapi.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.gamequest.server.gameapi.session.GameQuestSessionRegistry;
import systems.zlink.samples.gamequest.shared.contracts.Messages;

/** Receives projection notifications from QuestMission and pushes them to the bound stream session. */
@ZLinkHandlerGroup("gameapi")
public final class NotifyQuestProgressHandler
    implements ZLinkRequestHandler<Messages.NotifyQuestProgressReq, Messages.NotifyQuestProgressRes> {
    private final GameQuestSessionRegistry registry;

    public NotifyQuestProgressHandler(GameQuestSessionRegistry registry) {
        this.registry = registry;
    }

    @Override
    public Messages.NotifyQuestProgressRes handle(
        Messages.NotifyQuestProgressReq request, ZLinkRequestContext context) {
        return new Messages.NotifyQuestProgressRes(registry.notify(request));
    }
}

File: framework/languages/java/samples/kotlin/GameQuest/Server/GameApi/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/server/gameapi/handlers/NotifyQuestProgressHandler.kt
package systems.zlink.samples.kotlin.gamequest.server.gameapi.handlers

import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.gamequest.server.gameapi.session.GameQuestSessionRegistry
import systems.zlink.samples.kotlin.gamequest.shared.contracts.NotifyQuestProgressReq
import systems.zlink.samples.kotlin.gamequest.shared.contracts.NotifyQuestProgressRes

/** Receives projection notifications from QuestMission and pushes them to the bound stream session. */
@ZLinkHandlerGroup("gameapi")
class NotifyQuestProgressHandler(
    private val registry: GameQuestSessionRegistry,
) : ZLinkSuspendingRequestHandler<NotifyQuestProgressReq, NotifyQuestProgressRes> {
    override suspend fun handle(request: NotifyQuestProgressReq, context: ZLinkRequestContext): NotifyQuestProgressRes =
        NotifyQuestProgressRes(registry.notify(request))
}
